File: Creational/Singleton.py
# from typing import Optional
import threading

# # hungry mode
# class Singleton:
#     instance = None

#     def __init__(self) -> None:
#         if Singleton.instance is not None:
#             raise Exception("Only one instance of Singleton class is allowed.")
#         else:
#             Singleton.instance = self


# # lazy mode
# class Singleton:
#     __instance: Optional['Singleton'] = None # 双下划线表示私有变量

#     @staticmethod
#     def getInstance():
#         if Singleton.__instance is None:
#             Singleton.__instance = Singleton()
#         return Singleton.__instance


# lazy mode with Double-Checked Locking
class Singleton:
    __instance = None
    __lock = threading.Lock()

    @classmethod
    def getInstance(cls):
        if cls.__instance is None:
            with cls.__lock:  # 保证线程安全，同一时刻内只有一个线程可以创建单例
                if (
                    cls.__instance is None
                ):  # 二次检查，防止等待锁的过程中，其他线程已经创建了实例
                    cls.__instance = Singleton()
        return cls.__instance

File: Creational/Singleton.js
import { createInterface } from "readline";
// const { createInterface } = require("readline")

const rl = createInterface({
  input: process.stdin,
  output: process.stdout,
});

const PRIVATE_KEY = Symbol();

// lazy mode
class ShoppingCartManager {
  static #instance; // private
  #products = [];

  // 构造函数使用 PRIVATE_KEY 设置为私有
  constructor(key) {
    if (key !== PRIVATE_KEY) {
      throw new Error("Only one instance of Singleton class is allowed.");
    }
  }

  static getInstance() {
    if (!ShoppingCartManager.#instance) {
      ShoppingCartManager.#instance = new ShoppingCartManager(PRIVATE_KEY);
    }
    return ShoppingCartManager.#instance;
  }

  add(name, quantity) {
    this.#products.push({ name, quantity });
  }

  viewCart() {
    if (this.#products.length === 0) {
      return;
    }
    this.#products.forEach((product) => {
      console.log(`${product.name} ${product.quantity}`);
    });
  }
}

const cartList = [];
rl.on("line", (line) => {
  cartList.push(line.split(" "));
});

rl.on("pause", () => {
  // 暂停时才处理输出逻辑，写在函数外会直接执行
  cart = ShoppingCartManager.getInstance();
  for (let x of cartList) {
    cart.add(...x);
  }
  cart.viewCart();
  rl.close();
});
